test-install.js

```diff
@@ -9,32 +9,44 @@ const debug = require('debug')('node-chromium');
 
 const utils = require('./utils');
 const config = require('./config');
+const install = require('./install');
 
-const install = async () => {
-    await require('./install');
-};
-
-test.before(t => {
+test.beforeEach(t => {
     // Deleting output folder
     const outPath = config.BIN_OUT_PATH;
     debug(`Deleting output folder: [${outPath}]`);
 
     if (fs.existsSync(outPath)) {
         rimraf.sync(outPath);
     }
+
+    // Ensure a consistent, known environment for each test
+    process.env.CHROMIUM_DOWNLOAD_HOST = '';
+    process.env.CHROMIUM_REVISION = '';
+
     t.pass();
 });
 
 test.serial('Canary Test', t => {
     t.pass();
 });
 
-test('Before Install Process', t => {
+test.serial('Before Install Process', t => {
     const binPath = utils.getOsChromiumBinPath();
     t.false(fs.existsSync(binPath), `Chromium binary is found in: [${binPath}]`);
 });
 
-test('Chromium Install', async t => {
+test.serial('Chromium Install', async t => {
+    await install();
+
+    const binPath = utils.getOsChromiumBinPath();
+    const isExists = fs.existsSync(binPath);
+    t.true(isExists, `Chromium binary is not found in: [${binPath}]`);
+});
+
+test.serial('Chromium Install from Mirror', async t => {
+    process.env.CHROMIUM_DOWNLOAD_HOST = 'https://npm.taobao.org/mirrors/chromium-browser-snapshots/';
+    process.env.CHROMIUM_REVISION = '737027';
     await install();
```

Comment:
Can we extract these 4 instructions to separate function not to have duplicate code (something like `installChromiumAndVerify`)? What do you think? Does it make sense here?

Reply:
I have refactored the tests to use a helper. Also you asked what I think so here goes 😃 ...

I have learned over the years to avoid helpers in unit tests and to prefer cut and paste reuse 😱 I'm not saying "never, ever use helpers in unit tests", it's a rule of thumb. I have noticed other senior software devs doing this and have had conversations with several about it, they tend to agree.

There are a few reasons for this, here's the general idea... What tends to happen on larger projects with lots of tests is that your helper needs to take more and more params, here's an invented example:

```js

test('Test foo', function () {
    fooTestHelper();
});

test('Test foo with bar', function () {
    fooTestHelper('bar');
});

test('Test foo with bar and X', function () {
    fooTestHelper('bar', 'X');
});

test('Test foo without bar and X', function () {
    fooTestHelper(null, 'X');
});

test('Test foo with bar and X and Y', function () {
    fooTestHelper('foo', 'X', 'Y');
});

test('Test foo without bar and X and Y', function () {
    fooTestHelper(null, 'X', 'Y');
});

test('Test foo without bar without X and Y', function () {
    fooTestHelper(null, null, 'Y');
});

// etc, etc .... on it goes and on it grows
```

It could be cleaned up a bit by refactoring the helper to use DTOs BUT that's not the main problem.

The **main problem is that the helper ends up becoming extremely complex** because it has to handle so many different scenarios. Instead of **each test standing alone** and being completely **easy to understand** that behemoth test helper which made sense during the accretion phase is now daunting and unintelligible.  When a test fails the last thing you want is extra complexity. 

Another major issue is that **adding new tests becomes painful** - again, you have to understand an overly complex test helper, possibly it doesn't quite do what you want, now you have to modify the helper **and affect every single test!** The stakes are super high, any mistakes here and you have broken the entire test suite! Instead of just understanding the one piece of functionality you are changing and testing you have to understand every piece of functionality in every permutation to make sure you haven't invalidated existing tests. Oh the horror.

I'm sure you can find people who disagree, ultimately it's a trade-off and depends a lot on the situation.

I think for now the helper is fine but I can easily imagine it growing if you start adding configuration params.

Reply:
BTW if I have convinced you then I'm happy to revert the tests to get rid of the helper. 
If I was the maintainer I would not want a test helper it but it's your call.

Reply:
Yes, I agree with you. I don't like code duplication (because of the need of changing the code in multiple places) but here, I don't like the function itself neither (I prefer functions with single responsibility but here it's not as easy to make one). That's why I asked your opinion and I do agree that it might be better to leave it as it is (with duplicates).

Reply:
I actually have removed the "mirror install test" - I have made a more detailed comment about this below.